src/nvim/option.c

```diff
@@ -726,11 +726,173 @@ void ex_set(exarg_T *eap)
+{
+  varnumber_T value;
+
+  // ":set opt!": invert
+  // ":set opt&": reset to default value
+  // ":set opt<": reset to global value
+  if (nextchar == '!') {
+    value = *(int *)(varp) ^ 1;
+  } else if (nextchar == '&') {
+    value = (int)(intptr_t)options[opt_idx].def_val;
+  } else if (nextchar == '<') {
+    // For 'autoread' -1 means to use global value.
+    if ((int *)varp == &curbuf->b_p_ar && opt_flags == OPT_LOCAL) {
+      value = -1;
+    } else {
+      value = *(int *)get_varp_scope(&(options[opt_idx]), OPT_GLOBAL);
+    }
+  } else {
+    if (prefix == 2) {                  // inv
+      value = *(int *)varp ^ 1;
+    } else {
+      value = prefix;
+    }
+  }
+
+  *errmsg = set_bool_option(opt_idx, (char_u *)varp, (int)value, opt_flags);
+}
+
+static void do_set_num(int opt_idx, int opt_flags, char **argp, int nextchar, const set_op_T op,
+                       const char *varp, char *errbuf, size_t errbuflen, char **errmsg)
+{
+  varnumber_T value;
+  char *arg = *argp;
+
+  // Different ways to set a number option:
+  // &            set to default value
+  // <            set to global value
+  // <xx>         accept special key codes for 'wildchar'
+  // c            accept any non-digit for 'wildchar'
+  // [-]0-9       set number
+  // other        error
+  arg++;
+  if (nextchar == '&') {
+    value = (long)(intptr_t)options[opt_idx].def_val;
+  } else if (nextchar == '<') {
+    // For 'undolevels' NO_LOCAL_UNDOLEVEL means to
+    // use the global value.
+    if ((long *)varp == &curbuf->b_p_ul && opt_flags == OPT_LOCAL) {
+      value = NO_LOCAL_UNDOLEVEL;
+    } else {
+      value = *(long *)get_varp_scope(&(options[opt_idx]), OPT_GLOBAL);
+    }
+  } else if (((long *)varp == &p_wc
+              || (long *)varp == &p_wcm)
+             && (*arg == '<'
+                 || *arg == '^'
+                 || (*arg != NUL && (!arg[1] || ascii_iswhite(arg[1]))
+                     && !ascii_isdigit(*arg)))) {
+    value = string_to_key(arg);
+    if (value == 0 && (long *)varp != &p_wcm) {
+      *errmsg = e_invarg;
+      return;
+    }
+  } else if (*arg == '-' || ascii_isdigit(*arg)) {
+    int i;
+    // Allow negative, octal and hex numbers.
+    vim_str2nr(arg, NULL, &i, STR2NR_ALL, &value, NULL, 0, true);
+    if (i == 0 || (arg[i] != NUL && !ascii_iswhite(arg[i]))) {
+      *errmsg = e_number_required_after_equal;
+      return;
+    }
+  } else {
+    *errmsg = e_number_required_after_equal;
+    return;
+  }
+
+  if (op == OP_ADDING) {
+    value = *(long *)varp + value;
+  }
+  if (op == OP_PREPENDING) {
+    value = *(long *)varp * value;
+  }
+  if (op == OP_REMOVING) {
+    value = *(long *)varp - value;
+  }
+  *errmsg = set_num_option(opt_idx, (char_u *)varp, (long)value,
+                           errbuf, errbuflen, opt_flags);
+}
+
+// Handle some special cases with string option values
+static void munge_string_opt_val(char **varp, char **oldval, char **const origval,
```

Comment:
I've got a hard time finding a reliable translation of "munge", is it like "misc" or |pot-pourri" ?